src/apps/intel1g/intel1g.lua

```diff
@@ -0,0 +1,684 @@
+MDICNFG   0x00E04 -            RW MDI Configuration
+RCTL      0x00100 -            RW RX Control
+SW_FW_SYNC 0x05b5c -           RW Software Firmware Synchronization
+TCTL      0x00400 -            RW TX Control
+TCTL_EXT  0x00400 -            RW Extended TX Control
+]],
+   txq = [[
+TDBAL  0xe000 +0x40*0..7 RW Tx Descriptor Base Low
+TDBAH  0xe004 +0x40*0..7 RW Tx Descriptor Base High
+TDLEN  0xe008 +0x40*0..7 RW Tx Descriptor Ring Length
+TDH    0xe010 +0x40*0..7 RO Tx Descriptor Head
+TDT    0xe018 +0x40*0..7 RW Tx Descriptor Tail
+TXDCTL 0xe028 +0x40*0..7 RW Tx Descriptor Control Queue
+TXCTL  0xe014 +0x40*0..7 RW Tx DCA CTRL Register Queue
+]]
+}
+reg.i210 = {
+   inherit = "1000BaseX",
+   singleton = [[
+EEMNGCTL  0x12030 -            RW Manageability EEPROM-Mode Control Register
+EEC       0x12010 -            RW EEPROM-Mode Control Register
+]]
+}
+reg.i350 = {
+   inherit = "1000BaseX",
+   singleton = [[
+EEMNGCTL  0x01010 -            RW Manageability EEPROM-Mode Control Register
+EEC       0x00010 -            RW EEPROM-Mode Control Register
+]]
+}
+
+local Intel = { }
+function Intel:new (arg)
+   local conf = config.parse_app_arg(arg)
+   local self = setmetatable({
+      r = {},
+      pciaddress = conf.pciaddr,
+      ndesc = conf.ndescriptors or 1024,
+      txq = conf.txq,
+      rxq = conf.rxq,
+      rssseed = conf.rssseed or 314159
+   }, {__index = self})
+   local deviceInfo = pci.device_info(self.pciaddress)
+   assert(deviceInfo.vendor == '0x8086', "unsupported nic")
+   local models = {}
+   models["0x1521"] = "i350"
+   models["0x1533"] = "i210"
+   models["0x157b"] = "i210"
+   models["0x10fb"] = "82599ES"
+   local ringSize = {}
+   ringSize["i350"] = 8
+   ringSize["i210"] = 4
+   ringSize["82599ES"] = 128
+
+   self.model    = models[deviceInfo.device]
+   assert(self.model, "Unsupported Intel NIC")
+   self.ringSize = ringSize[self.model]
+
+   -- Setup device access
+   self.base, self.fd = pci.map_pci_memory_unlocked(self.pciaddress, 0)
+   self.master = self.fd:flock("ex, nb")
+
+   self:load_registers(self.model)
+
+   self:init()
+   self.fd:flock("sh")
+   self:init_tx_q()
+   self:init_rx_q()
+   return self
+end
+
+function Intel:disable_interrupts ()
+   self.r.EIMC(0xffffffff)
+end
+function Intel:init_rx_q ()
+   if not self.rxq then return end
+   assert((self.rxq >=0) and (self.rxq < self.ringSize),
+   "rxqueue must be in 0.." .. self.ringSize-1 .. " for " .. self.model)
+   assert((self.ndesc %128) ==0,
+   "ndesc must be a multiple of 128 (for Rx only)")	-- see 7.1.4.5
+
+   self.rxpackets = ffi.new("struct packet *[?]", self.ndesc)
+   self.rdh = 0
+   self.rdt = 0
+   -- setup 4.5.9
+   local rxdesc_t = ffi.typeof([[
+   struct {
+      uint64_t address;
+      uint16_t length, cksum;
+      uint8_t status, errors;
+      uint16_t vlan;
+   } __attribute__((packed))
+   ]])
+   assert(ffi.sizeof(rxdesc_t),
+   "sizeof(rxdesc_t)= ".. ffi.sizeof(rxdesc_t) .. ", but must be 16 Byte")
```

Comment:
Should be `assert(ffi.sizeof(rxdesc_t) == 16, ...)`? Not sure how much sense the assertion makes though, in which cases would it not hold true?